storage.go

```diff
@@ -4,33 +4,140 @@ import (
 	"context"
 	"io"
 
+	ipfs "github.com/ipfs/go-ipfs-api"
+
+	"github.com/beyondstorage/go-storage/v4/pkg/iowrap"
+	"github.com/beyondstorage/go-storage/v4/services"
 	. "github.com/beyondstorage/go-storage/v4/types"
 )
 
+// The src of `ipfs files cp` supports both `IPFS-path` and `MFS-path`
+// After `s.getAbsPath(src)`, if the absolute path matches `IPFS-path`, it will take precedence
+// This means that if the `workDir` is `/ipfs/`, there is a high probability that an error will be returned
+// See https://github.com/beyondstorage/specs/pull/134#discussion_r663594807 for more details
+func (s *Storage) copy(ctx context.Context, src string, dst string, opt pairStorageCopy) (err error) {
+	return s.ipfs.FilesCp(ctx, s.getAbsPath(src), s.getAbsPath(dst))
+}
+
 func (s *Storage) create(path string, opt pairStorageCreate) (o *Object) {
-	panic("not implemented")
+	if opt.HasObjectMode && opt.ObjectMode.IsDir() {
+		path += "/"
+		o = NewObject(s, true)
+		o.Mode = ModeDir
+	} else {
+		o = NewObject(s, false)
+		o.Mode = ModeRead
+	}
+	o.ID = s.getAbsPath(path)
+	o.Path = path
+	return o
 }
 
+// GSP-46: Idempotent Storager Delete Operation
+// ref: https://github.com/beyondstorage/specs/blob/master/rfcs/46-idempotent-delete.md
 func (s *Storage) delete(ctx context.Context, path string, opt pairStorageDelete) (err error) {
-	panic("not implemented")
+	err = s.ipfs.FilesRm(ctx, s.getAbsPath(path), true)
+	return
 }
 
 func (s *Storage) list(ctx context.Context, path string, opt pairStorageList) (oi *ObjectIterator, err error) {
-	panic("not implemented")
+	rp := s.getAbsPath(path)
+
+	var nextFn NextObjectFunc
+	switch {
+	case opt.ListMode.IsPart():
+	case opt.ListMode.IsDir():
+		nextFn = func(ctx context.Context, page *ObjectPage) error {
+			dir, err := s.ipfs.FilesLs(ctx, rp, ipfs.FilesLs.Stat(true))
```

Comment:
Does `ipfs.FilesLs.Stat(true)` consumes a lot?

We have [object lazy stat](https://beyondstorage.io/docs/go-storage/internal/object-lazy-stat) feature, so we don't need to stat while in `List`.

Reply:
Here the `Stat` corresponds to the param `long`, if not specified, only the `name` can be obtained.

Reply:
If follow the `lazy stat` strategy, should I modify the function as follows:

```Golang
nextFn := func(ctx context.Context, page *ObjectPage) error {
	dir, err := s.ipfs.FilesLs(ctx, rp, ipfs.FilesLs.Stat(false))
	if err != nil {
		return err
	}
	for _, f := range dir {
		o := NewObject(s, false)
		o.Path = f.Name
		page.Data = append(page.Data, o)
	}
	return IterateDone
}
```

I want to know which properties in `Object` are necessary to be set here, and if only `path` is set will there be a problem?

Reply:
Only `id`, `path`, and `mode` are required here. Other fields will be fetched if needed.

(I will add a clear comment for it.)

Reply:
comments added in https://github.com/beyondstorage/go-storage/pull/614/files